cli/internal/cacheitem/cacheitem.go

```diff
@@ -0,0 +1,76 @@
+// Package cacheitem is an abstraction over the creation and restoration of a cache
+package cacheitem
+
+import (
+	"archive/tar"
+	"compress/gzip"
+	"crypto/sha512"
+	"errors"
+	"hash"
+	"io"
+	"log"
+	"os"
+	"reflect"
+	"sync"
+
+	"github.com/vercel/turborepo/cli/internal/turbopath"
+)
+
+var (
+	errMissingSymlinkTarget = errors.New("symlink restoration is delayed")
+	errCycleDetected        = errors.New("links in the cache are cyclic")
+	errTraversal            = errors.New("tar attempts to write outside of directory")
+	errNameMalformed        = errors.New("file name is malformed")
+	errNameWindowsUnsafe    = errors.New("file name is not Windows-safe")
+	errUnsupportedFileType  = errors.New("attempted to restore unsupported file type")
+)
+
+// CacheItem is a `tar` utility with a little bit extra.
+type CacheItem struct {
+	// Path is the location on disk for the CacheItem.
+	Path turbopath.AbsoluteSystemPath
+	// Anchor is the position on disk at which the CacheItem will be restored.
+	Anchor turbopath.AbsoluteSystemPath
+
+	// For creation.
+	once   sync.Once
+	sha    hash.Hash
+	tw     *tar.Writer
+	gzw    *gzip.Writer
+	handle *os.File
+}
+
+// Close any open pipes
+func (ci *CacheItem) Close() error {
+	// Close from the beginning of the pipe to the end.
+	closers := []io.Closer{ci.tw, ci.gzw, ci.handle}
+
+	for _, closer := range closers {
+		// Skip the things which may not exist in this particular instance.
+		if reflect.ValueOf(closer).IsZero() {
```

Comment:
In what scenarios would this happen? And is this different than a nil check?

Reply:
The `nil` check panics, so I switched to `reflect` in order to make the check. Not sure if there is a better solution?

Reply:
Of the readers and writers, only the `file` implementation of `io.Closer` throws an error on multiple calls to `Close()`; I think it's reasonable to insulate ourselves from that error.

https://go.dev/play/p/e8n7IcLYHf6
https://github.com/golang/go/blob/master/src/archive/tar/writer.go#L463-L483
https://github.com/golang/go/blob/master/src/compress/flate/deflate.go#L630-L651

The error pattern I used is explicitly borrowed from the implementations of gzip/tar in stdlib, `isClosed` is not a state we can confirm, but having an error _is_ a state we can confirm. `AddFile` is the only method which needs to interact with `Close()`, though `Open` and `Restore` are actually redundant (right now).